knimin/handlers/barcode_util.py

```diff
@@ -258,34 +163,135 @@ def update_ag_barcode(self, barcode):
+                    barcode=barcode, email_type=None,
+                    barcode_info=barcode_details,
+                    proj_barcode_info=ag_details, msgs=None,
+                    currentuser=self.current_user)
+
+    @authenticated
+    def post(self):
+        barcode = self.get_argument('barcode')
+        postmark_date = self.get_argument('postmark_date', None)
+        scan_date = self.get_argument('scan_date', None)
+        biomass_remaining_value = self.get_argument('biomass_remaining_value',
+                                                    None)
+        sequencing_status = self.get_argument('sequencing_status', None)
+        obsolete_status = self.get_argument('obsolete_status', None)
+        projects = set(self.get_arguments('project'))
+        sent_date = self.get_argument('sent_date', None)
+        login_user = self.get_argument('login_user',
+                                       'American Gut participant')
+        send_mail = self.get_argument('send_mail', None)
+        login_email = self.get_argument('login_email', '')
+        other_text = self.get_argument('other_text', None)
+        email_type = self.get_argument('email_type', None)
+        sample_time = self.get_argument('sample_time', None)
+        sample_date = self.get_argument('sample_date', None)
+        # now we collect data and update based on forms
+        # first update general barcode info
+        # Set to non to make sure no conflicts with new date typing in DB
+        if not postmark_date:
+            postmark_date = None
+        if not scan_date:
+            scan_date = None
         try:
-            db.updateAKB(barcode, moldy, overloaded, other,
-                         self.get_argument('other_text', None),
-                         sent_date)
-            msg3 = ("Barcode %s AG info was sucessfully updated" % barcode)
+            db.updateBarcodeStatus('Received',
+                                   postmark_date,
+                                   scan_date, barcode,
+                                   biomass_remaining_value,
+                                   sequencing_status,
+                                   obsolete_status)
+            msg1 = "Barcode %s general details updated" % barcode
         except:
-            msg3 = ("Barcode %s AG update failed!!!" % barcode)
+            msg1 = "Barcode %s general details failed" % barcode
 
-        return msg2, msg3
+        msg2 = msg3 = msg4 = None
+        exisiting_proj, parent_project = db.getBarcodeProjType(
+            barcode)
+        exisiting_proj = set(exisiting_proj.split(','))
+        if exisiting_proj != projects:
+            try:
+                add_projects = projects.difference(exisiting_proj)
+                rem_projects = exisiting_proj.difference(projects)
+                db.setBarcodeProjects(barcode, add_projects, rem_projects)
+                msg4 = "Project successfully changed"
+            except:
```

Comment:
It catches any and all errors so that the interface can reflect an error occurred.